lib/puma/cluster.rb

```diff
@@ -337,8 +351,7 @@ def stop
     def stop_blocked
       @status = :stop if @status == :run
       wakeup!
-      @control.stop(true) if @control
-      Process.waitall
+      sleep 0.2 until @status == :halt
```

Comment:
? Purpose?

Reply:
`Process.waitall` no longer completes here, because during a re-fork one of the parent's children is the promoted process which doesn't exit but instead becomes the new parent.

Instead, the halt status is set after the promotion process is complete and all other worker-children have terminated.